working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+# Shared Memory Multithreading for Dart
+
+This proposal tries to address two long standing gaps which separate Dart from other more low-level languages:
+
+* Dart developers should be able to utilize multicore capabilities without hitting limitations of the isolate model.
+
+* Interoperability between Dart and native platforms (C/C++, Objective-C/Swift, Java/Kotlin) requires aligning concurrency models between Dart world and native world. The misalignment is not an issue when invoking simple native synchronous APIs from Dart, but it becomes a blocker when:
+
+  * working with APIs pinned to a specific thread (e.g. `@MainActor` APIs in Swift or UI thread APIs on Android) 
+
+  * dealing with APIs which want to call _back_ from native into Dart on an arbitrary thread.
+
+  * using Dart as a container for shared business logic - which can be invoked on an arbitrary thread by surrounding native application. 
+
+  Native code does not _understand_ the concept of isolates. 
+
+The biggest challenge with introducing shared memory multithreading into Dart is avoiding breaking existing code, which is written under the assumption of single-threaded access. We would like to avoid just outright breaking existing code in subtle ways. Consider for a moment a library which is maintains a global cache internally:  
+
+```     dart
+library foo;
+
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId;
+final _cache = <int, C>{};
+
+C makeObject() {
+  map[nextId] = C(nextId);
+  nextId++;
+  return map[nextId];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the same code becomes thread _unsafe_ in the presence of shared memory multithreading.
+
+## Shareable Data
+
+We propose to extend Dart with a shared memory multithreading but provide a clear type-based delineation between two concurrency worlds. **Type of an instance should provide a clear indication whether its fields can be concurrently mutated by another thread.** We are going to call such types _shareable_ and assume that the language provides some way to opt-in into this for classes:
+
+```  dart
+/// Mark class sharable by implementing a [Shareable]
```

Comment:
What if we made all classes with only sharable fields implicitly sharable? Then `implements Sharable` would just be a way of asking the compiler to throw an error if the class *isn't* implicitly sharable.

Reply:
Could work but hinders evolution.
If absence of unsendable fields makes a type sendable, adding an unsendable field is a breaking change.
Even if you didn't *know*, and didn't care, that your class was sendable.

That's a bad pattern, and why we are very strict about requiring constants to be marked `const` and only promoting instance variables that are declared in specific safe ways, because otherwise a small change that looks fine *locally* might break the derived feature for someone else.

Generally, you won't get an affordance from my code that I haven't explicitly opted into, because otherwise I might change something that removes the affordance, which I never cared for, or knew about, myself.

So making things sharable just because we *can* goes against that feature design principle.

Reply:
> That makes it hard for `Null` to implement it.

Why couldn't we simply define `Null` as `final class Null implements Shareable {}`?